sled-agent/src/bootstrap/early_networking.rs

```diff
@@ -38,59 +47,238 @@ pub enum EarlyNetworkSetupError {
+                    &format!("http://[{}]:{}", addr, MGS_PORT),
+                    self.log.new(o!("component" => "MgsClient")),
+                );
+
+                info!(
+                    self.log, "Querying MGS to determine switch location";
+                    "addr" => %addr,
+                );
+                let switch_slot = mgs_client
+                    .sp_local_switch_id()
+                    .await
+                    .with_context(|| format!("Failed to query MGS at {addr}"))?
+                    .into_inner()
+                    .slot;
+
+                match switch_slot {
+                    0 => Ok((SwitchLocation::Switch0, addr)),
+                    1 => Ok((SwitchLocation::Switch1, addr)),
+                    _ => Err(anyhow!(
+                        "Nonsense switch slot returned by MGS at \
+                         {addr}: {switch_slot}"
+                    )),
+                }
+            });
+
+        let mut switch_location_map = HashMap::new();
+        for mgs_query_result in future::join_all(mgs_query_futures).await {
+            match mgs_query_result {
+                Ok((location, addr)) => {
+                    info!(self.log, "Found {location:?} at {addr}");
+                    switch_location_map.insert(location, addr);
+                }
+                Err(err) => {
+                    warn!(self.log, "{err}");
+                }
+            }
+        }
+
+        // Are we done? If we found 2 switch locations we're done (we only have
+        // at most 2 switches); we could also be done if we only have 1 switch,
+        // if internal DNS reported there is only one switch to find.
+        match switch_location_map.len() {
+            2 => Ok(LookupSwitchZoneAddrsResult::TotalSuccess(
+                switch_location_map,
+            )),
+            1 => {
+                if switch_zone_addrs.len() == 1 {
+                    Ok(LookupSwitchZoneAddrsResult::TotalSuccess(
+                        switch_location_map,
+                    ))
+                } else {
+                    Ok(LookupSwitchZoneAddrsResult::PartialSuccess(
+                        switch_location_map,
+                    ))
+                }
+            }
+            0 => Err(BackoffError::transient(
+                "No switch locations found".to_string(),
+            )),
+            _ => unreachable!(
+                "HashMap keyed on `SwitchLocation` can have at most 2 entries"
+            ),
+        }
     }
 
-    // Initialize the rack network and return the boundary switch addresses to
-    // be injected into zone requests
-    pub async fn init_rack_network(
+    // Initialize a single switch via DPD.
+    pub async fn init_switch_config(
         &mut self,
         rack_network_config: &RackNetworkConfig,
-        switch_mgmt_addrs: &HashMap<SwitchLocation, Ipv6Addr>,
-    ) -> Result<HashSet<Ipv6Addr>, EarlyNetworkSetupError> {
+        switch_ip: Ipv6Addr,
```

Comment:
`switch_zone_underlay_ip` ?

Reply:
I did some various renaming to be more explicit (and added more comments) in 56f1e4f2, 592ba19a, and 152ddb86